File: beetle/api/task.py
from typing import Literal

from fastapi import APIRouter, Body

from beetle.models import Task
from beetle.constants import TaskStatus


router = APIRouter()


@router.post("/create")
async def create(
    src: str = Body(),
    dst: str = Body(),
    interval: int = Body(ge=300),
    cleanup: bool = Body(),
):
    task = await Task.filter(src=src, dst=dst).first()
    if task:
        return {"error": "task already exists"}
    task = await Task.create(
        src=src,
        dst=dst,
        interval=interval,
        cleanup=cleanup,
        status=TaskStatus.running,
    )
    return task


@router.post("/update")
async def update(
    id: int = Body(),
    src: str = Body(),
    dst: str = Body(),
    interval: int = Body(le=60),
    cleanup: bool = Body(),
    status: Literal[TaskStatus.running, TaskStatus.stopped] = Body(),
):
    rows = await Task.filter(id=id).update(
        src=src,
        dst=dst,
        interval=interval,
        cleanup=cleanup,
        status=status,
    )
    return rows


@router.post("/delete")
async def delete(id: int = Body(embed=True)):
    rows = await Task.filter(id=id).delete()
    return rows


@router.post("/reads")
async def reads(constraints: dict = Body(default={})):
    tasks = await Task.filter(**constraints).all()
    return tasks

File: web/src/api/task.ts
import request from './request'

export type TaskType = {
  id: number
  src: string
  dst: string
  interval: number
  cleanup: boolean
  status: 'running' | 'stopped'
  created_at: string
  updated_at: string
  _create?: boolean
}

export async function reads() {
  return request<TaskType[]>('/task/reads', {
    method: 'POST',
  })
}

export async function remove(id: number) {
  return request('/task/delete', {
    method: 'POST',
    body: JSON.stringify({ id: id }),
  })
}

export async function create(payload: {
  src: string
  dst: string
  cleanup: boolean
  interval: number
  status: string
}) {
  return request<TaskType>('/task/create', {
    method: 'POST',
    body: JSON.stringify(payload),
  })
}

export async function update(payload: {
  id: number
  src: string
  dst: string
  cleanup: boolean
  interval: number
  status: string
}) {
  return request<number>('/task/update', {
    method: 'POST',
    body: JSON.stringify(payload),
  })
}

export const Task = {
  reads: reads,
  remove: remove,
  create: create,
  update: update,
}
